applications/pionix_chargebridge/src/utilities/filesystem.cpp

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2025 Pionix GmbH and Contributors to EVerest
+#include <charge_bridge/utilities/filesystem.hpp>
+
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <random>
+
+namespace charge_bridge::filesystem_utils {
+
+bool read_from_file_partial(const fs::path& file_path, const std::size_t byte_count, std::string& out_data) {
+    try {
+        if (fs::is_regular_file(file_path)) {
+            std::ifstream file(file_path, std::ios::binary);
+
+            if (file.is_open()) {
+                std::vector<char> buffer(byte_count);
+                file.read(buffer.data(), byte_count);
+
+                std::size_t read_bytes = file.gcount();
+
+                if (read_bytes == byte_count) {
+                    out_data.assign(buffer.data(), read_bytes);
+                    return true;
+                }
+            }
+        }
+    } catch (const std::exception& e) {
+        return false;
+    }
+
+    return false;
+}
+
+bool read_from_file(const fs::path& file_path, std::string& out_data) {
+    try {
+        if (fs::is_regular_file(file_path)) {
+            std::ifstream file(file_path, std::ios::binary);
+
+            if (file.is_open()) {
+                out_data = std::string((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+                return true;
+            }
+        }
+    } catch (const std::exception& e) {
+        return false;
+    }
+
+    return false;
+}
+
+bool process_file(const fs::path& file_path, size_t buffer_size,
+                  std::function<bool(const std::vector<std::uint8_t>&, bool last_chunk)>&& func) {
+    std::ifstream file(file_path, std::ios::binary);
+
+    return process_file(file, buffer_size, std::move(func));
+}
+
+bool process_file(std::ifstream& file, size_t buffer_size,
+                  std::function<bool(const std::vector<std::uint8_t>&, bool last_chunk)>&& func) {
+    if (!file) {
+        return false;
+    }
+
+    std::vector<std::uint8_t> buffer(buffer_size);
+    bool interupted = false;
+
+    while (file.read(reinterpret_cast<char*>(buffer.data()), buffer_size)) {
+        interupted = func(buffer, false);
+
+        if (interupted) {
+            break;
+        }
+    }
+
+    // Process the remaining bytes
+    if (interupted == false) {
+        size_t remaining = file.gcount();
+
+        // Keep only remaining elements
+        buffer.resize(remaining);
+        func(buffer, true);
+    }
+
+    return true;
+}
+
+// Returns true on success, fills `hdr`, and sets `image_offset` to the byte
+// position (from start of file) where the firmware image begins.
+bool read_crypt_signed_header(const fs::path& path, CryptSignedHeader& hdr, std::uint32_t& image_offset) {
```

Comment:
Do we really need to use output parameters here? Could we not just return a struct?